tikv-client/src/main/java/com/pingcap/tikv/operation/iterator/ScanIterator.java

```diff
@@ -17,59 +17,52 @@
 
 import static java.util.Objects.requireNonNull;
 
-import com.google.common.collect.Range;
 import com.google.protobuf.ByteString;
 import com.pingcap.tikv.TiSession;
 import com.pingcap.tikv.exception.TiClientInternalException;
-import com.pingcap.tikv.kvproto.Coprocessor.KeyRange;
+import com.pingcap.tikv.key.Key;
 import com.pingcap.tikv.kvproto.Kvrpcpb;
 import com.pingcap.tikv.kvproto.Metapb;
 import com.pingcap.tikv.region.RegionManager;
 import com.pingcap.tikv.region.RegionStoreClient;
 import com.pingcap.tikv.region.TiRegion;
 import com.pingcap.tikv.util.BackOffer;
 import com.pingcap.tikv.util.ConcreteBackOffer;
-import com.pingcap.tikv.util.KeyRangeUtils;
 import com.pingcap.tikv.util.Pair;
-import com.pingcap.tikv.key.Key;
 import java.util.Iterator;
 import java.util.List;
-import java.util.NoSuchElementException;
 
 public class ScanIterator implements Iterator<Kvrpcpb.KvPair> {
-  private final Range<Key> scanRange;
-  private final int batchSize;
   protected final TiSession session;
   private final RegionManager regionCache;
   protected final long version;
 
   private List<Kvrpcpb.KvPair> currentCache;
   protected ByteString startKey;
   protected int index = -1;
-  private boolean endOfRegion = false;
+  private boolean endOfScan = false;
 
   public ScanIterator(
       ByteString startKey,
-      int batchSize,
-      KeyRange range,
       TiSession session,
-      RegionManager rm,
       long version) {
     this.startKey = requireNonNull(startKey, "start key is null");
-    this.batchSize = batchSize;
-    if (range == null) {
-      this.scanRange = Range.all();
-    } else {
-      this.scanRange = KeyRangeUtils.makeRange(range.getStart(), range.getEnd());
+    if (startKey.isEmpty()) {
+      throw new IllegalArgumentException("start key cannot be empty");
```

Comment:
is this logic reachable?

Reply:
If previous region is the right most region, then its EndKey will be next scan's startKey which is empty.